Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -0,0 +1,1191 @@
+#align quadratic_form QuadraticForm
+
+namespace QuadraticForm
+
+section FunLike
+
+variable [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable {Q Q' : QuadraticForm R M}
+
+instance funLike : FunLike (QuadraticForm R M) M fun _ => R where
+  coe := toFun
+  coe_injective' x y h := by cases x; cases y; congr
+#align quadratic_form.fun_like QuadraticForm.funLike
+
+/-- Helper instance for when there's too many metavariables to apply
+`FunLike.hasCoeToFun` directly. -/
+instance : CoeFun (QuadraticForm R M) fun _ => M → R :=
+  ⟨FunLike.coe⟩
+
+variable (Q)
+
+/-- The `simp` normal form for a quadratic form is `coeFn`, not `toFun`. -/
+@[simp]
+theorem toFun_eq_coe : Q.toFun = ⇑Q :=
+  rfl
+#align quadratic_form.to_fun_eq_coe QuadraticForm.toFun_eq_coe
+
+-- this must come after the coe_to_fun definition
+initialize_simps_projections QuadraticForm (toFun → apply)
+
+variable {Q}
+
+@[ext]
+theorem ext (H : ∀ x : M, Q x = Q' x) : Q = Q' :=
+  FunLike.ext _ _ H
+#align quadratic_form.ext QuadraticForm.ext
+
+theorem congr_fun (h : Q = Q') (x : M) : Q x = Q' x :=
+  FunLike.congr_fun h _
+#align quadratic_form.congr_fun QuadraticForm.congr_fun
+
+theorem ext_iff : Q = Q' ↔ ∀ x, Q x = Q' x :=
+  FunLike.ext_iff
+#align quadratic_form.ext_iff QuadraticForm.ext_iff
+
+/-- Copy of a `QuadraticForm` with a new `toFun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy (Q : QuadraticForm R M) (Q' : M → R) (h : Q' = ⇑Q) : QuadraticForm R M where
+  toFun := Q'
+  toFun_smul := h.symm ▸ Q.toFun_smul
+  exists_companion' := h.symm ▸ Q.exists_companion'
+#align quadratic_form.copy QuadraticForm.copy
+
+@[simp]
+theorem coe_copy (Q : QuadraticForm R M) (Q' : M → R) (h : Q' = ⇑Q) : ⇑(Q.copy Q' h) = Q' :=
+  rfl
+#align quadratic_form.coe_copy QuadraticForm.coe_copy
+
+theorem copy_eq (Q : QuadraticForm R M) (Q' : M → R) (h : Q' = ⇑Q) : Q.copy Q' h = Q :=
+  FunLike.ext' h
+#align quadratic_form.copy_eq QuadraticForm.copy_eq
+
+end FunLike
+
+section Semiring
+
+variable [Semiring R] [AddCommMonoid M] [Module R M]
+
+variable (Q : QuadraticForm R M)
+
+theorem map_smul (a : R) (x : M) : Q (a • x) = a * a * Q x :=
+  Q.toFun_smul a x
+#align quadratic_form.map_smul QuadraticForm.map_smul
+
+theorem exists_companion : ∃ B : BilinForm R M, ∀ x y, Q (x + y) = Q x + Q y + B x y :=
+  Q.exists_companion'
+#align quadratic_form.exists_companion QuadraticForm.exists_companion
+
+theorem map_add_add_add_map (x y z : M) :
+    Q (x + y + z) + (Q x + Q y + Q z) = Q (x + y) + Q (y + z) + Q (z + x) := by
+  obtain ⟨B, h⟩ := Q.exists_companion
+  rw [add_comm z x]
+  simp [h]
+  abel
+#align quadratic_form.map_add_add_add_map QuadraticForm.map_add_add_add_map
+
+theorem map_add_self (x : M) : Q (x + x) = 4 * Q x := by
+  rw [← one_smul R x, ← add_smul, map_smul]
+  norm_num
+#align quadratic_form.map_add_self QuadraticForm.map_add_self
+
+-- porting note: removed @[simp]
```

Comment:
I assume this is because of the `simpNF` linter?
```suggestion
-- porting note: removed @[simp] because it is superseded by `ZeroHomClass.map_zero`
```